File: articles/implementation-scaling-design-thinking/c/implementation_value_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ITEMS 128
#define MAX_LINE 512
#define MAX_NAME 180

typedef struct {
    char name[MAX_NAME];
    double adoption_readiness;
    double operational_fit;
    double durability;
    double governance_readiness;
    double equity_readiness;
    double financial_sustainability;
    double operational_risk;
    double governance_risk;
    double technical_risk;
    double equity_risk;
    double financial_risk;
    double composite_risk;
    double implementation_value;
} Intervention;

double compute_composite_risk(const Intervention *x) {
    return 0.25 * x->operational_risk +
           0.22 * x->governance_risk +
           0.18 * x->technical_risk +
           0.22 * x->equity_risk +
           0.13 * x->financial_risk;
}

double compute_implementation_value(const Intervention *x) {
    return 0.20 * x->adoption_readiness +
           0.18 * x->operational_fit +
           0.18 * x->durability +
           0.15 * x->governance_readiness +
           0.14 * x->equity_readiness +
           0.10 * x->financial_sustainability -
           0.05 * x->composite_risk;
}

int compare_desc(const void *a, const void *b) {
    const Intervention *ia = (const Intervention *)a;
    const Intervention *ib = (const Intervention *)b;

    if (ia->implementation_value < ib->implementation_value) return 1;
    if (ia->implementation_value > ib->implementation_value) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/implementation_portfolio_raw.csv";
    FILE *file = fopen(path, "r");

    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    Intervention interventions[MAX_ITEMS];
    char line[MAX_LINE];
    int count = 0;

    fgets(line, sizeof(line), file); /* header */

    while (fgets(line, sizeof(line), file) && count < MAX_ITEMS) {
        char *token = strtok(line, ",");
        if (!token) continue;

        strncpy(interventions[count].name, token, MAX_NAME - 1);
        interventions[count].name[MAX_NAME - 1] = '\0';

        strtok(NULL, ","); /* intervention_type */
        strtok(NULL, ","); /* implementation_stage */

        token = strtok(NULL, ",");
        interventions[count].adoption_readiness = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].operational_fit = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].durability = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].governance_readiness = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].equity_readiness = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].financial_sustainability = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].operational_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].governance_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].technical_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].equity_risk = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].financial_risk = token ? atof(token) : 0.0;

        interventions[count].composite_risk = compute_composite_risk(&interventions[count]);
        interventions[count].implementation_value = compute_implementation_value(&interventions[count]);

        count++;
    }

    fclose(file);

    qsort(interventions, count, sizeof(Intervention), compare_desc);

    printf("rank,intervention,implementation_value,composite_risk\n");
    for (int i = 0; i < count; i++) {
        printf("%d,%s,%.6f,%.6f\n", i + 1, interventions[i].name, interventions[i].implementation_value, interventions[i].composite_risk);
    }

    return 0;
}

File: articles/implementation-scaling-design-thinking/go/implementation_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Intervention struct {
	Name                    string
	AdoptionReadiness       float64
	OperationalFit          float64
	Durability              float64
	GovernanceReadiness     float64
	EquityReadiness         float64
	FinancialSustainability float64
	OperationalRisk         float64
	GovernanceRisk          float64
	TechnicalRisk           float64
	EquityRisk              float64
	FinancialRisk           float64
	CompositeRisk           float64
	ImplementationValue     float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func compositeRisk(x Intervention) float64 {
	return 0.25*x.OperationalRisk +
		0.22*x.GovernanceRisk +
		0.18*x.TechnicalRisk +
		0.22*x.EquityRisk +
		0.13*x.FinancialRisk
}

func implementationValue(x Intervention) float64 {
	return 0.20*x.AdoptionReadiness +
		0.18*x.OperationalFit +
		0.18*x.Durability +
		0.15*x.GovernanceReadiness +
		0.14*x.EquityReadiness +
		0.10*x.FinancialSustainability -
		0.05*x.CompositeRisk
}

func loadPortfolio(path string) []Intervention {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []Intervention

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 14 {
			log.Fatalf("malformed row: %v", row)
		}

		x := Intervention{
			Name:                    row[0],
			AdoptionReadiness:       parseFloat(row[3]),
			OperationalFit:          parseFloat(row[4]),
			Durability:              parseFloat(row[5]),
			GovernanceReadiness:     parseFloat(row[6]),
			EquityReadiness:         parseFloat(row[7]),
			FinancialSustainability: parseFloat(row[8]),
			OperationalRisk:         parseFloat(row[9]),
			GovernanceRisk:          parseFloat(row[10]),
			TechnicalRisk:           parseFloat(row[11]),
			EquityRisk:              parseFloat(row[12]),
			FinancialRisk:           parseFloat(row[13]),
		}
		x.CompositeRisk = compositeRisk(x)
		x.ImplementationValue = implementationValue(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/implementation_portfolio_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadPortfolio(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].ImplementationValue > items[j].ImplementationValue
	})

	fmt.Println("rank,intervention,implementation_value,composite_risk")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.ImplementationValue, x.CompositeRisk)
	}
}
